grant-project/zcash-grant-system
# Proposal contribution matching

Closes #88.

I welcome feedback on the UX, I like the ribbon. But I'm not sure I love the `CampaignBlock` "x2", might be too subtle. But maybe it's ok? 

I suspect we might want to have more control/definition on this feature's behavior, but I think it wise to have a final pass at that once refunds, payout & staking basics are also present.

### What this does
- adds `contribution_matching` float to `Proposal` model
- derives funded amount on the frontend using `contributionMatching`
- limits its value to 1 (on) or 0 (off), allows arbitrary multipliers in the future
- admin: adds ability to set matching status 
- frontend: ribbon visual cue for proposal list cards with matching turned on
- frontend: subtle callout in `CampaignBlock` for matching

### What this doesn't do
- limit when a proposal can have its matching state set
- limit the amount a user can donate based on multiplier (user can contribute the full remaining amount if they want)
- display proposal accounting for admin to easily find out how much is owed via matching

### Testing
- upgrade your database, this should add the column and set the `contribution_matching` default to '0' (via `server_default`)
- use admin to set any proposal to "matching" (switch under delete button)
- make sure the proposal is live
- make sure the card has the ribbon callout
- make sure the proposal campaign block has the "x2" on the "Funding" line, should have a tooltip
- contributions to the proposal should be doubled (on list, detail and profile > created)
- turn on matching for a contribution with > 50% actual funding, make sure it is listed as fully funded, and the amount funded is equal to the target and not greater than the target.

## Screens
![screen shot 2019-01-28 at 2 33 38 pm](https://user-images.githubusercontent.com/11430954/51866346-6440e680-230e-11e9-8a5e-88d75bd2fd1c.png)

![screen shot 2019-01-28 at 2 41 09 pm](https://user-images.githubusercontent.com/11430954/51866358-699e3100-230e-11e9-9e01-c7ee9aa684e7.png)

![screen shot 2019-01-28 at 2 37 17 pm](https://user-images.githubusercontent.com/11430954/51866366-6e62e500-230e-11e9-994a-5f4480c86e5a.png)
